data/base.go

```diff
@@ -0,0 +1,256 @@
+
+	base := Base{
+		_ID:              bson.NewObjectId(),
+		ID:               bson.NewObjectId().Hex(),
+		UserID:           ToString(UserIDField, datum[UserIDField], errs),
+		DeviceID:         ToString(deviceIDField, datum[deviceIDField], errs),
+		UploadID:         ToString(uploadIDField, datum[uploadIDField], errs),
+		Time:             ToString(timeField, datum[timeField], errs),
+		Type:             ToString(typeField, datum[typeField], errs),
+		Payload:          datum[payloadField],
+		ConversionOffset: ToInt(conversionOffsetField, datum[conversionOffsetField], errs),
+		TimezoneOffset:   ToInt(timezoneOffsetField, datum[timezoneOffsetField], errs),
+		ClockDriftOffset: ToInt(clockDriftOffsetField, datum[clockDriftOffsetField], errs),
+		DeviceTime:       ToString(deviceTimeField, datum[deviceTimeField], errs),
+		Annotations:      ToArray(annotationsField, datum[annotationsField], errs),
+		Internal: Internal{
+			GroupID:       datum[GroupIDField].(string),
+			ActiveFlag:    true,
+			SchemaVersion: 1, //TODO: configured ??
+			CreatedTime:   time.Now().Format(time.RFC3339),
+		},
+	}
+
+	validator.SetErrorReasons(validationFailureReasons).Struct(base, errs)
+
+	return base
+}
+
+var validationFailureReasons = validate.ErrorReasons{
+	timeStringTag:     timeStrValidationMsg,
+	timeZoneOffsetTag: fmt.Sprintf("TimezoneOffset needs to be in minutes and greater than %d and less than %d", tzValidationLowerLimit, tzValidationUpperLimit),
+}
+
+func PayloadValidator(v *valid.Validate, topStruct reflect.Value, currentStructOrField reflect.Value, field reflect.Value, fieldType reflect.Type, fieldKind reflect.Kind, param string) bool {
+	//a place holder for more through validation
+	if _, ok := field.Interface().(interface{}); ok {
+		return true
+	}
+	return false
+}
+
+func AnnotationsValidator(v *valid.Validate, topStruct reflect.Value, currentStructOrField reflect.Value, field reflect.Value, fieldType reflect.Type, fieldKind reflect.Kind, param string) bool {
+	//a place holder for more through validation
+	if _, ok := field.Interface().([]interface{}); ok {
+		return true
+	}
+	return false
+}
+
+func TimezoneOffsetValidator(v *valid.Validate, topStruct reflect.Value, currentStructOrField reflect.Value, field reflect.Value, fieldType reflect.Type, fieldKind reflect.Kind, param string) bool {
+	if offset, ok := field.Interface().(int); ok {
+		if offset >= tzValidationLowerLimit && offset <= tzValidationUpperLimit {
+			return true
+		}
+	}
+	return false
+}
+
+func TimeObjectValidator(v *valid.Validate, topStruct reflect.Value, currentStructOrField reflect.Value, field reflect.Value, fieldType reflect.Type, fieldKind reflect.Kind, param string) bool {
+	if timeObject, ok := field.Interface().(time.Time); ok {
+		return isTimeObjectValid(timeObject)
+	}
+	return false
+}
+
+func isTimeObjectValid(timeObject time.Time) bool {
+	return !timeObject.IsZero() && timeObject.Before(time.Now())
```

Comment:
Technically, a time object is valid even if it is after now. We might want to clarify this in the naming, "PastTimeObjectValidator"?